src/Random.mo

```diff
@@ -0,0 +1,188 @@
+/// A module for obtaining randomness on the Internet Computer (IC).
+///
+/// This module provides the fundamentals for user abstractions to build on.
+///
+/// Dealing with randomness on a deterministic computing platform, such
+/// as the IC, is intricate. Some basic rules need to be followed by the
+/// user of this module to obtain (and maintain) the benefits of crypto-
+/// graphic randomness:
+/// - cryptographic entropy (randomness source) is only obtainable
+///   asyncronously in discrete chunks of 256 bits (32-byte sized `Blob`s)
+/// - all bets must be closed *before* entropy is being asked for in
+///   order to decide them
+/// - this implies that the same entropy (i.e. `Blob`) - or surplus entropy
+///   not utilised yet - cannot be used for a new round of bets without
+///   losing the cryptographic guarantees.
+///
+/// Concretely, the below class `Finite`, as well as the
+/// `*From` methods risk the carrying-over of state from previous rounds.
+/// These are provided for performance (and convenience) reasons, and need
+/// special care when used. Similar caveats apply for user-defined (pseudo)
+/// random number generators.
+
+import Prim "mo:prim";
+import P "Prelude";
+import I "Iter"
+
+module {
+
+  /// Drawing from a finite supply of entropy, `Finite` provides
+  /// methods to obtain random values. When the entropy is used up,
+  /// `null` is returned. Otherwise the outcomes' distributions are
+  /// stated for each method. The uniformity of outcomes is
+  /// guaranteed only when the supplied entropy is originally obtained
```

Comment:
Okay, so in other words, it is always a mistake to mix uses of different functions on a single Finite object. But wouldn't that suggest a different interface, then? It would be more adequate to put them in separate classes, so that this can't happen. Moreover, if you do, then each such class could be modelled naturally as an iterator. Like:
```
class Coin(entropy : Blob) {
  ...
  public func next() : ?Bool { ... }
};

class Range(n : Nat, entropy : Blob) {
  ...
  public func next() : ?Nat { ... }
};

class Binomial(n : Nat8, entropy : Blob) {
  ...
  public func next() : ?Nat8 { ... }
};
```
This is safer, clearer, and you can even use it nicely in for loops.

Reply:
> always a mistake to mix uses of the different methods on a single Finite object

I don't see how you come to that conclusion. There are 256 bits of entropy (when seeded from `blob()`) and how you spend it is your thing. Users can trivially build the classes you suggest on top of `Finite`. E.g.
``` swift
class Coin(entropy : Blob) : I.Iter<Bool> {
  let fin = Finite(entropy)
  public func next() : ?Bool = fin.coin()
};
```
(modulo syntax errors).

Reply:
Hm, is the outcome of Finite.range still uniformly distributed when it is interleaved with calls to coin?

Reply:
It is. The iterator used for `coin` grabs the entropy byte wise and serves it out as bits. Nothing is ever reused.

Reply:
Is was wondering about bits _not_ being used, but I think see why that shouldn't matter. Thanks.

Reply:
Exactly. One can ignore entropy or use it. Also reorder. But you are not allowed to reuse, as it would create non-orthogonalities (entanglement, bayesian coupling).